### ISSUE TITLE
Memory Leak in BiVariableMap when Sharing Variables is enabled

### ISSUE THREAD

**Issue by matthias-fratz-bsz** [jruby/jruby#8527]
**Environment Information**
  
Provide at least:
* JRuby version range: at least 9.4.3.0 through commit 9d63c228
* Operating system: Linux 6.1.0-26-amd64 SMP PREEMPT_DYNAMIC Debian 6.1.112-1 (2024-09-30) x86_64 GNU/Linux

**Test Case**

[InstanceVariableMemoryLeak.java](https://github.com/user-attachments/files/18098929/InstanceVariableMemoryLeak.java.txt)

This code creates many instances of a Ruby object, each of which contains several instance variables. No references to those Ruby objects are held anywhere in the example code. The code does a GC run to try and get rid of them, then drops a heap dump (because I like heap dumps).

**Expected Behavior**

Because no reference to the Ruby objects is held anywhere in user code, they are garbage-collected during that GC run. The heap dump is fairly boring as a result. Note that when the Sharing Variables feature is explicitly disabled, that's exactly what happens.

**Actual Behavior**

(with Variable Sharing enabled, which is the default)

The JRuby objects are never garbage-collected and neither are their instance variables. The heap dump shows that `BiVariableMap` contains over 100k `BiVariable` objects:

![Eclipse MAT screenshot](https://github.com/user-attachments/assets/f9ea3d34-85c7-48f5-b88a-45761a76784d)

([According to the docs](https://github.com/jruby/jruby/wiki/RedBridge#disabling-sharing-variables), it should be possible to disable Variable Sharing with `System.setProperty("org.jruby.embed.sharing.variables", "false");` or `-Dorg.jruby.embed.sharing.variables=false`... but that doesn't seem to work with the test case. `container.setAttribute(AttributeName.SHARING_VARIABLES, false);` does work. That problem is somewhat unrelated, and maybe it's just a documentation bug and system properties aren't meant to become container attributes anyway, but I figured I should mention it.)

**my analysis so far**

All those Ruby instance variables are stored into a `BiVariableMap` object for access from Java (observable with a breakpoint or print() in `BiVariableMap.update()`). There is no mechanism for removing those once the Ruby object has been garbage-collected; at least I cannot find any in the code.

Each `BiVariable` object has a `receiver` member which references the Ruby object. Thus the Ruby object itself never becomes unreachable, and is thus also never garbage-collected.

The variable entries in BiVariableMap are created from `EmbedEvalUnitImpl.run()` if Sharing Variables is enabled:

```
    final BiVariableMap vars = container.getVarMap();
    final boolean sharing_variables = isSharingVariables(container);
...
    if (sharing_variables) {
        vars.retrieve(ret);
    }
```

They are actually created in `InstanceVariable.updateInstanceVar()`, called via `BiVariableMap.retrieve()`, `VariableIntercetpro.retrieve()` and `InstanceVariable.update()`. `InstanceVariable.updateInstanceVar()` first tries to update an existing `BiVariable` in `BiVariableMap`, but then creates a new `InstanceVariable` object when that variable doesn't yet exist.

I strongly suspect that the `receiver` reference in `BiVariable` could be removed. Looking at `BiVariableMap`, the only way to get hold of a `BiVariable` with `receiver != getTopSelf()` is:

- calling `getVariable(RubyObject, String)` and passing a reference to that Ruby object in the first place
- calling `getVariables()` (unfortunately)

All the `Map` interfaces actually ignore the `receiver` object, sometimes in inconsistent ways. For example, if there is an instance variable called `foo`, `containsKey("foo") == true` but `get("foo") == null` because `get` checks the receiver object where `containsKey` doesn't.

That is, it should be possible to store the values of instance variables in the `RubyObject` instead, where they are garbage-collected along with that Ruby object. Or lazily create the `InstanceVariable` objects in `BiVariableMap` only when they are actually set, and the caller can be expected to remove them after use.

The big issue with this approach that I cannot figure out, is how to implement the `getVariables()` method. It is public, so there is no way of knowing whether any code out there relies on it to access instance variables. There are more specific accessors (`get(Object, Object)`, `getVariable(RubyObject, String)`, `put(Object, String, Object)` etc) so there is no need to use `getVariables()` for that... but existing code might use it anyway :(

**Comment by headius** [comment]
This also should be fixed for 9.4.12.0 along with #8422. Since we don't have a patch for this year, it makes sense to wait on merging the fix (#8483) until the next release.

But it is high priority for 9.4.12.0.

**Comment by matthias-fratz-bsz** [comment]
@headius After thinking about this for a bit, I think this is impossible to fix this memory leak without breaking some existing use of the embedding interface. The worst possible (mis)use I can think of is:

1. run some ruby code that creates and returns an object with instance variable `@foo`
2. don't bother to hold a reference to the returned object in Java
3. use `BiVariableMap.getVariables()` and scan the returned list to find the `InstanceVariable` for `@foo`
4. retrieve the value of `@foo` (or do something else with it)

This obviously relies on the ruby object (along with its `@foo` instance variable) not being garbage-collected despite being effectively unreachable. I'd consider such usage to be buggy to begin with, but it is currently guaranteed to work: All `InstanceVariable`s are held in the `BiVariableMap`, so neither `@foo` nor its receiver object can be garbage collected.

In order to fix the memory leak, we'd need to change the logic so that both the ruby object and `@foo` become eligible for garbage-collection after step 2. This means step 3 and beyond might find that the object and its instance variables have disappeared. I don't think that it's a big problem, but I can totally imagine "creative" developers doing that kind of stuff.

In fact, the code very much looks like it is intended to be used that way. `BiVariableMap.retrieve(RubyObject)` is called in 3 cases:

1. in `EmbedEvalUnitImpl.run()`, on the returned object (exactly case 2 above)
2. in `EmbedRubyObjectAdapterImpl.doInvokeMethod()`, on the receiver object on which the method is presumably being called
3. in `EmbedRubyObjectAdapterImpl.doInvokeSuper()`, on the receiver object

_So it doesn't add all instance variables, just the ones in objects returned from Ruby to Java._ (I guess that makes sense, but I missed it.)

Looking at it this way, the "sharing variables" feature totally makes sense: It's a convenient way of accessing Ruby variables from Java, without having to worry about the lifecycle of Ruby objects. It does mean that the Java code becomes responsible for cleaning up the instance variables of essentially any object that Java code ever touched, but if that's documented it wouldn't be that hard to do. In most cases where I'd use such a simplified interface, it would probably be a straight-forward case where I can just zap everything in the `BiVariableMap` when I'm done with whatever I'm using Ruby to do.

However, if that is what the "sharing variables" feature does, I strongly believe that the it should not be enabled by default. Java programmers will expect a returned JRuby object to be garbage-collected like any other (Java) object. If I'm calling methods on a Ruby object instance, I sure wouldn't expect that I will have to clean up its instance variables afterwards. Judging by all the memory leak issues here, neither do most other developers, apparently.

I couldn't find any documentation on what exactly the "sharing variables" feature actually does. The only docs I readily found is [the RedBridge wiki page](https://github.com/jruby/jruby/wiki/RedBridge#user-content-Disabling_Sharing_Variables), and that only really says that it's on by default and that "When the feature is off, slight performance improvement will be expected." (By the way: In my testcase above, that performance improvement is absolutely massive.)

If there is documentation about the "sharing variables" feature, I believe it should clearly state that Java has to clean up Ruby instance variables afterwards, and also what "afterwards" means, ie. after which methods / operations. Also, that documentation should be linked from that RedBridge wiki page. If there is no such documentation, I've probably just drawn the short straw on who has to write some...

So... my suggested fix is to:

- turn off variable sharing by default, because something needing explicit cleanup shouldn't be the default
- clearly document what it does and that one has to clean up after it
- mention in the release notes how to re-enable it, for those people whose code is broken by the change... probably also mentioning the need for cleanup again

**Comment by matthias-fratz-bsz** [comment]
Note: I haven't fully checked whether the "sharing variables" feature affects stuff other than instance variables. The code in `VariableInterceptor.inject()` looks like it might also do something with local variable values... but then I also don't really understand what a `DynamicScope` is and what it does here.

If "sharing variables" does handle injection of local variables as well, then it may be harder to just turn it off because more people might use that. But if that's what it does, it also looks like it's a poor idea to change the `BiVariableMap` between parsing and evaulating a script...? In any case, I'll have to give that a closer look next week.

Feel free to point me at any documentation about that feature!

**Comment by matthias-fratz-bsz** [comment]
Guess I should just have checked with a piece of code: Yes, sharing variables also handles injection of local variables. Example:

```java
final ScriptingContainer scriptEngine = new ScriptingContainer(LocalContextScope.SINGLETHREAD,
		LocalVariableBehavior.PERSISTENT);
scriptEngine.setAttribute(AttributeName.SHARING_VARIABLES, true);
scriptEngine.getVarMap().put("call", "meow");
System.out.println(scriptEngine.runScriptlet("call"));
scriptEngine.runScriptlet("call='woof'");
System.out.println(scriptEngine.runScriptlet("call"));
System.out.println(scriptEngine.getVarMap().get("call"));
```

This prints "meow woof woof", proving that `BiVariableMap` does provide bidirectional access to local variables as well. That is, I'm taking back the suggestion of disabling it by default; I don't think it is very practical.

@headius Looking at the code a bit more and getting a feel for how it can be used: I think `BiVariableMap` should just delegate to `IRubyObject.get/setInstanceVariable` for all instance variables (everything that starts with an `@` if I'm understanding Ruby correctly). That way, `BiVariableMap` never stores instance variables and they just go away along with the `RubyObject` they belong to. I think I can create a pull request for that solution.

This means instance variables won't show up in `BiVariableMap.getVariables()` any more. Hopefully nobody relies on that.

If the receiver is not an `IRubyObject`, we could throw `IllegalArgumentException` because it doesn't make sense. Or keep the current behavior when attempting to inject an instance variable into a Java object using that interface: Silently ignore the request, without reporting any errors.

**Comment by headius** [comment]
Revisiting this...

I have re-pushed your other PR to fix the *ContextProvider leaks, in #8969. Assuming that is green I think it's ok to merge in for 9.4.14.0 and fix part of the leaks you found.

For this one... your thought to just push the variables directly into the object makes some sense, but it's obviously a fairly visible change we would probably not want to do in a 9.4.x maintenance release. Perhaps we can come up with a better model for 10.0.x, which is still being adopted and stabilizing? And probably we should guard it behind an option, so we'd have the new instance-variable based version disabled by default for now and enable it later in 10.1 or so?

In any case, this is turning out to be too big a change to push into a 9.4.x release, so I'm moving it to 10.0.3 for re-examination.

**Comment by headius** [comment]
Limited time and resources... we were unable to get back to this for 10.0.3.0. There's larger questions about how to handle the sharing here, so it's getting bumped again.

**Comment by headius** [comment]
We're trying to do a quick flip of 10.0.4.0 so no new progress here.

**Comment by matthias-fratz-bsz** [comment]
@headius First of all, sorry for essentially disappearing on that issue. My priorities at work are currently elsewhere now that the leak no longer affects Cantaloupe. I may have to work on this from a non work account instead. (Or if anyone else feels like implementing this then obviously I'm very fine with that.)

When looking at the `BiVariableMap` there's 3 things I'd like to change:
1. Make all `Map` methods comply with the `Map` interface, and be consistent with each other. This means the `Map` methods would only allow access to the instance variables of the top self object only, plus all other non-instance variables like globals etc. (Instance variable `@foo` can exist in multiple objects after all, and the `Map` interface cannot represent that.) I think that's all they can do right now anyway, but I'd have to confirm in detail.
2. Mark all methods that take an `Object receiver` parameter as `@Deprecated`. These are redundant with `RubyObject.getInstanceVariables()` and then using the various methods of the returned `InstanceVariable` interface. `BiVariableMap` doesn't provide a complete set of these in the first place, e.g. there is no `keySet(Object receiver)` that would do the equivalent of `InstanceVariables.getInstanceVariableNameList()`.
3. Provide the current behavior of `getVariables()` as a fallback for code that still needs to access instance variables of an object, but for some reason doesn't have a reference to that object. (I don't think such code has any excuse for not having a reference to that object: It's always the one it just called a method on, or the object that was returned by a method. But it's a backwards-incompatible change nevertheless.) I'd guard it behind a config option `SHARING_INSTANCE_VARIABLES` which is disabled by default and would document both the potential memory leak and when to clean up variables from `BiVariableMap` when using it. With `SHARING_INSTANCE_VARIABLES` disabled, `getVariables()` would probably have to throw an exception... or maybe it's better to remove the method entirely and provide a replacement instead that clearly documents that it needs `SHARING_INSTANCE_VARIABLES` enabled to do anything useful?

Unfortunately that's a lot of ideas that I'd need to try out before really making a pull request. And I don't really have time to do that right now :(

**Comment by matthias-fratz-bsz** [comment]
So... I did unexpectedly find some time to look into this some more. And I need to prepend that list with a step zero, mostly as a "note to self" but I might as well share it:

- Make `BiVariableMap` and the `org.jruby.embed.variable.*Variable` classes actually use the right ruby runtime (`Ruby` object). Probably also remove the `receiver` reference entirely from variable types that don't use it, like `GlobalVariable`.

`BiVariableMap` is specific to a particular thread when using `ThreadSafeLocalContextProvider`. That is, every calling thread has its own `BiVariableMap`. It also has its own runtime, and thus also its own Top Self object. However, the `BiVariableMap` for a thread doesn't actually hold a reference to this runtime, but instead delegates to `LocalContextProvider.getRuntime()` – which then retrieves the runtime for the *calling thread*, not the runtime that actually belongs to the `BiVariableMap`.

What this means is that the behavior can become somewhat "interesting" when the `BiVariableMap` is shared across threads. (NB: Doing that is ill-advised because it breaks the thread safety `ThreadSafeLocalContextProvider` is meant to provide.) For example:

```
ScriptingContainer container = new ScriptingContainer(LocalContextScope.THREADSAFE,
		LocalVariableBehavior.TRANSIENT);
final BiVariableMap instance = container.getVarMap();
instance.put("animal", "dog");
Thread t = new Thread() {
	@Override
	public void run() {
		System.out.println(instance.containsKey("animal")); // prints true
		System.out.println(instance.get("animal")); // prints null
		instance.put("animal", "cat");
	}
};
t.start();
t.join();
System.out.println(instance.get("animal")); // prints "dog"
System.out.println(instance.size()); // prints 2
```

That happens because variables without an explicit receiver object are stored with a receiver equal to the Top Self object. And since the Top Self object depends on the runtime and thus differs between threads, the map now contains 2 variables with name `animal`: One for the parent thread and one for the child thread. It isn't supposed to work like that; each `BiVariableMap` should only contain the variables for "its own" thread with `THREADSAFE`.

IMHO it's not a big issue in itself, but I find it really hard to understand what exactly the code does there. And it isn't handled in an entirely consistent way either: Note that the example above uses local variables, not instance variables. Instance variables don't have a receiver object at all... but `BiVariableMap` threats them as if they were tied to an object anyway. In this case, they're associated with the Top Self object, but `BiVariableMap.put(Object, String, Object)` will happily create a local variable associated with any other object. As far as I can tell, injecting local variables then works properly. (Except that it'll probably pick a random one of the multiple values present in the map.)

But my point is: I think that code would profit from a bit of refactoring first.

**Comment by matthias-fratz-bsz** [comment]
@headius I have looked into the code some more and I keep finding more structures that don't make a lot of sense to me.

I would eventually like to make a pull request for a more streamlined `BiVariableMap` interface if such a thing is desirable. There's a lot of logic that copies variables between the Ruby runtime and `BiVariableMap` before and after evaluation, with provisions for doing the copy lazily... but only local variables actually need to be set before evaluation, and then restored (rescued?) after evaluation before they go out of scope. Everything else (globals, class variables, instance variables, constants and ARGV) can instead delegate to various calls on the Ruby runtime or a `RubyObject` receiver if and when a variable is actually accessed.

I now believe that having `ScriptingContainer.get()` and `ScriptingContainer.put()` methods do add significant value: They allow setting variables from Java without worrying too much about (or fully understandin) Ruby semantics. They don't look difficult to implement without all the copying.

Keeping the APIs involving `BiVariable` (`getVariable()`, `setVariable()`, `updateVariable()` / `update(name, BiVariable)` and `getVariables()`) is almost impossible. Those essentially assume that `BiVariableMap` holds a copy of every variable in a suitable `BiVariable` instance. But if that group of APIs is considered expendable, I could probably implement the rest in a much simpler, and memory-leak-free, way. (That is, including arbitrary access to instance variables.)

----

The other option would be the blunt fix of excluding instance variables from `BiVariableMap` by default. I drafted how that would look in #9331 – essentially just skipping `updateInstanceVar` unless `SHARING_INSTANCE_VARIABLES` is explicitly set to true. That would add even more complexity to an already complex piece of code, but it does fix the leak without involving any API changes.

**Comment by headius** [comment]
@matthias-fratz-bsz I think your approach makes some sense and this API is in dire need of a rethink. This obviously won't make it for 10.0.5.0 due out this week but I'm going to review your notes in #9331 and keep this in my queue.

This embedding API was designed many years ago and has vert few updates. Many users now just use the direct approach and construct an org.jruby.Ruby to use, but we still need to maintain compatibility with JSR-223 and at least the major parts of the existing "RedBridge" embedding API.

The original contributor, @yokolet, moved on from the project at least a decade ago, so I think it's up to us to figure out a way to fix up this API.

**Comment by headius** [comment]
Still in progress.
